### ISSUE TITLE
Files without a project

### ISSUE THREAD

**Issue by jrieken** [OmniSharp/omnisharp-roslyn#207]
In VSCode we encounter solo `cs`-files which don't belong to a project. While we OmniSharp cannot provide semantic features (IntelliSense, type check, etc) we could enable syntactical features which require a parse tree only (Formatting, outline, etc)

**Comment by nosami** [comment]
Heh... nice to hear you talk about VSCode on here :) 

We should be able to support this scenario already, by specifying the folder to the cs file. We do this in other editors already.

**Comment by filipw** [comment]
@nosami This doesn't work right now AFAIK because the project systems (asp.net 5 and msbuild) try to find projects in the folder (sln and project.json respectively) and if they don't find any projects, they just exit without populating workspace with anything.

scriptcs project system is the only one that discovers loose files and adds them to workspace but it's `csx` only

**Comment by nosami** [comment]
Ah, ok. I was getting confused with the old server which does work this way.

**Comment by jrieken** [comment]
@filipw - and you go with mscorlib which comes with desktop clr/mono right? So no happy path for core clr users? wondering if we could funnel in some default lib for this case...

**Comment by filipw** [comment]
yeah right now it just uses desktop/mono mscorlib. 
in fact, the whole csx thing is #ifdeffed against ASPNET50 too

I have started a separate repo so that all the same information could be retrieved from CSX files when omnisharp itself runs on coreclr https://github.com/filipw/scriptcs-dnx this is very rough outline now, but this approach should be replacing the embedded scriptcs dlls in the future

btw, one more thing, at the moment roslyn scripting doesn't support core clr, in fact it doesn't even work on mono, so scriptcs uses roslyn binaries on windows and mono.csharp on mono

but going back to the original topic, I think orphan `*.cs` files could be treated similarly to `*.csx` files and just added to workspace in a loose way; that would like you say require determining the runtime, plus some organizational decisions regarding how to code it in - would that be a new specialized project system that runs at the end or would that be part of msbuildprojectsystem for example, to basically gather all the loose files it didn't need as part of msbuild projects

**Comment by jrieken** [comment]
Agreed - orphan ```cs``` files should be handled in a similar fashion to ```csx``` files. We already have a little magic in place that associates an unknown file to the best matching project (https://github.com/OmniSharp/omnisharp-roslyn/blob/master/src/OmniSharp/Roslyn/BufferManager.cs#L80) but there is no fallback or default project. So, in the same spirit of OmniSharp v1 I would keep this in place and add an OrphanProject which takes all other files.

**Comment by akshita31** [comment]
@filipw I have been trying to add a OrphanProjectSystem for the orphan files. But the main problem I am running into is how to identify whether a particular document is an orphan document or has a project associated with it. I tried to invoke the GetDocument on the workspace and if a null is returned, then assuming that this is a orphan file. However, the problem with this approach is that the ProjectUpdate runs asynchronously for MSBuild Project system, hence the initialization of the orphan system is running before the projects have been added to the workspace. Any suggestions ?
